### PR TITLE
[Config] Consume GLM-5.2 DSA indexer skip/share fields

### PR DESCRIPTION
**Pull request by zrr1999**
### PR Category
Environment Adaptation

### PR Types
New features

### Description

This PR is superseded by [#1961](https://github.com/PaddlePaddle/PaddleFleet/pull/1961), which contains the complete GLM-5.2 implementation against develop. This PR is closed without merging; the implementation continues in the linked PR.

Map zai-org HuggingFace `config.json` DSA fields that develop still drops, and make `DSAttention` consume them so GLM-5.2 indexer skip/share is not load-only.

This is a third-party training stack aligned to the published HF DSA layout, not an official trainer.

Field map:

- `index_topk_freq` → `dsa_indexer_topk_freq`
- `index_skip_topk_offset` → `dsa_indexer_skip_topk_offset`
- `indexer_types` → `dsa_indexer_types`
- `index_share_for_mtp_iteration` → `dsa_index_share_for_mtp_iteration`

Behavior:

- Without mapping, `register_attributes` drops the keys and later layers cannot build the published indexer layout.
- `indexer_types==shared` skips indexer construction and reuses the previous `full` layer top-k. `index_share_for_mtp_iteration=True` reuses the last decoder top-k on MTP layers. GPT spec `layer_number` is 0-based and matches `indexer_types`.
- Rejecting non-default values would fail the published GLM-5.2 config (`index_topk_freq=4`, `shared` types).
- `__post_init__` requires `dsa_indexer_topk_freq>=1`, `dsa_indexer_skip_topk_offset>=0`, types in `{full,shared}` with length `num_hidden_layers` and `[0]==full`, and share-for-MTP only with MTP plus at least one decoder `full`.
- Defaults stay `freq=1 / offset=0 / types=None / share=False` (per-layer top-k). Existing jobs that do not pass these HF fields keep prior behavior.

Also covers recompute/RNG helpers and DSA unit coverage used by this layout. Accuracy kernels, native TP1 autograd, and FP32 UAC paths are not in this PR. This structural Draft is a dependency of the combined PaddleFleet tip, not a standalone training claim.

Pipeline construction validates the resolved global segment layout before allocating layers or entering collectives. Shared decoder and MTP indexers must reside with their full decoder producer in the same segment; unsupported cross-rank or interleaved-segment layouts raise a descriptive error at construction. Head-empty offsets, full-indexer layouts, and weight-only MTP remain supported.

Validation used these combined stack heads on H800/CUDA/bfloat16:

- [PaddleFleet `d5a5379f1263`](https://github.com/PaddlePaddle/PaddleFleet/commit/d5a5379f12637e4dcdd0d07e484b194f8ca0c5cf)
- [PaddleFormers `ffde408b986e`](https://github.com/PaddlePaddle/PaddleFormers/commit/ffde408b986ea811269218e861d11ae87897987d)
- [Megatron-LM `d8b684169735`](https://github.com/zrr1999/Megatron-LM/commit/d8b6841697350023f7b71a5792dd636d9cf63023)
- [ms-swift `91a3bb86d72f`](https://github.com/zrr1999/ms-swift/commit/91a3bb86d72fb8e301bb1a920b7815846c02820e)

| Native entrypoint | Configuration | Steps per framework | Result |
|---|---|---:|---|
| Formal training CLI | TP2 / PP2 / EP1 / SP, GBS1 | 100 | Strict loss, provenance and checkpoint checks passed; raw loss and 187 final tensors bitwise identical |
| GLM52 precision case | TP1 / PP2 / EP2, GBS2 | 100 | Downloaded model shards loaded by both frameworks; raw loss and 187 final tensors bitwise identical |

Both runs reproduce the corresponding pre-change loss bits and updated parameters. Formal invocation: `formal-pr-review-20260909-a-upload-n100-r1`; native case: `ci-glm52-upload-20260909-n100-r1`. The native case used the existing local dataset override; it is not a remote CI execution or a claim about the remote Minimax data cache.

Focused validation includes 34 DSA configuration/layout tests and 13 Trainer tests. The latter execute FP32 and BF16 AdamW save → load → next-step comparisons, check original state object/shape restoration after save failures, and exercise the model-config accuracy guard. Relevant repository format hooks passed.

The official-checkpoint subset is available as [PaddleFormers/GLM-5.2-BF16-minimal](https://git.aistudio.baidu.com/PaddleFormers/GLM-5.2-BF16-minimal.git) at `685a4b891d004a64bd2f1fef320deeffe162af02`. Independent remote download verified all 13 files; both weight shards preserve the original 187 tensor names, shapes, dtypes and bytes. Source: `zai-org/GLM-5.2@b4734de4facf877f85769a911abafc5283eab3d9`.

Remote CI still requires its model cache and matching companion builds. Known failures include review approval, a PaddleFleet wheel missing the deferred-token API, absent GLM4 cache metadata, ms-swift runner permissions/upstream formatting, and Minimax/GLM45 accuracy mismatches. Shared workflows and other model cases remain unchanged. Local GLM52 validation does not claim those checks passed.

Runtime remains Paddle `3.4.0.post20260808+733f3454aa0`, Torch `2.12.1+cu129`, cuBLAS `12.9.1.4`. Frozen source pins and the known Paddle wheel metadata dependency conflict remain separate from this implementation validation; canonical sealed acceptance is not claimed.

### 是否引起精度变化
否

Current PR head: `eece23290b46f2361b1a95eb03fddbfa8cd79b14`.

### CODE REVIEW

File: src/paddlefleet/transformer/dsa_attention.py [hunk 1]
```diff
@@ -1590,6 +1743,9 @@ def forward(
             _, topk_indices = self.indexer.forward(x, qr, indexer_float_mask)
             indexer_loss = None
 
+        if self.index_share and not self.skip_topk:
```
[the remark this answers is not in the document]
Comment (zrr1999) [reply src/paddlefleet/transformer/dsa_attention.py:1934]: Done. `b040d40b` 已统一 producer/consumer key：周期 skip 时 layer 0 发布 key 0，shared 层读同一 producer；官方 `['full', 'full', 'full', 'shared'] + index_share_for_mtp_iteration=True` 时 MTP 读 last-full key 2，不再读未发布的 key 3。holder 改挂在 process-local config 上，cloned mask 仍能看到同一份 top-k。测试见 `test_producer_and_consumer_holder_keys_match_for_legal_share_layouts` 与 `test_shared_consumer_reads_producer_holder_key`。

**Comment by zrr1999** [comment]
当前 head `b040d40b` 已处理 skip/share P1：producer/consumer holder key 对齐，holder 挂在 config 上，decoder 使用 logical layer index。请 @Paddle-Bot 对当前 head 重新评审。

File: src/paddlefleet/transformer/dsa_attention.py [hunk 2]
```diff
@@ -1393,6 +1393,126 @@ def track_indexer_metrics(
...
+    share_for_mtp_iteration = bool(
+        getattr(config, "dsa_index_share_for_mtp_iteration", False)
+    )
+    num_hidden_layers = getattr(config, "num_hidden_layers", 0) or 0
+    if is_mtp_layer:
+        # Official GLM-5.2 checkpoints still ship a full MTP indexer.
+        # Training honours share-for-MTP by skipping that indexer and
+        # reusing the last decoder layer's top-k, matching the HF field.
+        indexer_type = "shared" if share_for_mtp_iteration else "full"
+    elif indexer_types is not None and 0 <= layer_number < len(indexer_types):
```
[the remark this answers is not in the document]
Comment (zrr1999) [reply src/paddlefleet/transformer/dsa_attention.py]: Producer now publishes under `source_layer` (the logical producer id from `resolve_dsa_indexer_layout`), not physical `self.layer_number`. Skip/MTP consumers already look up that same key, including after `num_empty_layers_add_in_head`. Missing source keys still fail closed with `RuntimeError`; this PR does not reject PP>1+share at config time, and explicit pipeline tensor P2P of top-k remains out of scope.

Covered by:
- `test_head_empty_producer_publishes_logical_source_layer_key`
- `test_official_last_full_producer_key_matches_decoder_and_mtp`
- `test_skip_consumer_raises_when_source_key_is_missing`
- `test_shared_consumer_reads_producer_holder_key`

**Comment by zrr1999** [comment]
@Paddle-Bot 请对当前 head `525a3b65` 重新 review。本提交只把 `RecomputeStore` 改成对尚未发布该符号的 Paddle 做 ImportError 兼容（no-op stub），不改变 GLM-5.2 skip/share 结构语义。
